### ISSUE TITLE
playground for ts-generator

### ISSUE THREAD

**Issue by jechlin** [vojtechhabarta/typescript-generator#206]
Sometimes it's nice to play around with an class definition without having to build the whole thing, my build takes > 1 min.

I knocked up a prototype quickly - https://jechlin.github.io/ts-gen-pages/

Write groovy class(es), and press Convert. Conversion happens in AWS Lambda. 

It doesn't handle any compilation error at all at the moment! (Just experiment with changing field names). 

If not used for a while may take up to 4s to convert (I'll add a spinner). 

I think it should be able to handle java as an input, but only a single class I guess.

Basically there a ton of improvements I could make, but let me know if you think it could be useful, and any feedback.

Only tested in latest Chrome!

**Comment by vojtechhabarta** [comment]
Wow, that's great. How does it work? Does your lambda function call `new TypeScriptGenerator().generateTypeScript()` directly? And how do you compile groovy classes and put them to classpath? Maybe I should learn more about AWS Lambda 😉 

Actually I also planned to create web app but my goal was to develop TypeScript app which would help users with typescript-generator configuration. It would contain form with input fields, check boxes and radio buttons for important configuration parameters and it would show sample typescript-generator output for given configuration. It would also print configuration snippet for Maven and Gradle build file. I think typescript-generator configuration becomes more and more complex so I wanted some tool for users so that they can try different configuration options more quickly.

I intended to provide just sample output but having app which really runs typescript-generator is much better. And it could also include at least most important configuration parameters sometime in the future.

For me most important feature now would be support for Java. Each public Java class usually must be in separate file with the same name as class name but it can contain nested (static) classes so if there is a problem with multiple classes playground could have example with nested classes.

What are your plans with this app? Will it be accessible for users also in the future? If you want I will add link to README.

**Comment by jechlin** [comment]
> Does your lambda function call new TypeScriptGenerator().generateTypeScript() directly? And how do you compile groovy classes and put them to classpath? Maybe I should learn more about AWS Lambda

Yep. It's pretty simple in groovy to take some text and parse it to a class. 

AWS Lambda is not complex, it's just an easy and cheap way of running stuff on someone else's computer. The cost should be zero. I'll publish the source for the lambda function when I've improved it a bit. It's only about 10 lines.

> For me most important feature now would be support for Java.

I'll have a think about that.

> What are your plans with this app? Will it be accessible for users also in the future? If you want I will add link to README.

The intention was a simple way of previewing the generated output for different types of java constructs. As you say, I think it's important to be able to tweak the options to see how that affects the output.

So yes, I would definitely want it to be accessible to all users and potential users... just needs a bit more work first. 

One problem is how to update the version of the ts-gen-core jar. Would make sense for the lambda function code to be in your repo so you could deploy it when new versions are released. however then it would be your AWS key used, so you would incur any costs. Like I say I expect that to be zero - you get at least a million invocations per month for free.  I'll do some more work on it then we can come back to this problem.
cheers, jamie

**Comment by jechlin** [comment]
Hi Vojtech - 

I did a bit more work on it but I fear I may run out of time to do a proper job on it. 

New version is at https://jechlin.github.io/ts-gen-aws/, and the source is at https://github.com/jechlin/ts-gen-aws.

I had a go at adding options with some validation of the possible options, using npm module typescript-json-schema, and jsoneditor.

There is some basic error checking, but far from perfect. There is a ton of copy/pasting too. 

I'm happy for you to do any of these options below, or some combination.

1. Add issues and I'll work on it if I get time, no guarantees when though. For instance I didn't add java support - to be honest though the main difference between java/groovy for this stuff is you need square brackets not curlies for annotations that take arrays.
2. Link to it and hope that it's useful for someone, somewhere.
3. Incorporate the code into your repo, which will make deploying new versions easier.
4. Just ignore it, reimplement something yourself and take any of the ideas!

cheers, jamie

**Comment by vojtechhabarta** [comment]
Hi Jamie,
great progress!

Using JSON schema for options is really good idea! I was thinking about UI controls but JSON is easier to implement while still very usable for programmers. Have you also considered using [monaco](https://github.com/Microsoft/monaco-editor) editor (with [monaco-json](https://github.com/Microsoft/monaco-json) language plugin)? It is able to use JSON schema for validation and code completion, displaying documentation in hovers, handle comments in JSON, etc. So it should help users with adding parameters, selecting enum values and we could also have configuration template with parameters commented out (something similar to `tsc --init` output).

In any case adding link to [Configuration parameters](http://www.habarta.cz/typescript-generator/maven/typescript-generator-maven-plugin/generate-mojo.html) would be helpful for users I think.

My current plan is to upgrade typescript-generator to Java 8 so now I would just add a link.

Later we could move playground source to this repo (for example `playground` directory). Would it be possible to first just move the sources but could you still do deployment to AWS?

I don't want to add issues with requests on playground to you, I would condider it unfair since I personally don't have much time to work on it. But if you prefer to organize your work this way we can add them, or we can have just one issue (can be this) with list of tasks with checkboxes which we can add/modify/delete on the fly.

**Comment by jechlin** [comment]
> Have you also considered using monaco editor (with monaco-json language plugin

That's a good idea, I'm taking a look at that now.

>  I was thinking about UI controls

I thought about that too but frankly there are so many options I couldn't be bothered ;-) Not all of them are applicable though in this tool. I also tried feeding the schema into http://jeremydorn.com/json-editor/, but the generated form was just so big I didn't think it was useful, so then I switched to plain json. I don't really like the current validator/editor because it doesn't seem to provide any help when adding new keys. Possibly the schema is not quite right.

> Later we could move playground source to this repo (for example playground directory). Would it be possible to first just move the sources but could you still do deployment to AWS?

Yep, more than happy to do that.

**Comment by jechlin** [comment]
https://jechlin.github.io/ts-gen-aws/ - monaco editor is much better. Unfortunately json format doesn't allow comments... probably just have a set of sensible defaults for the most common options. Adding new keys is quite easy with this editor.

Back to my day job tomorrow... I'll try to improve it as and when.

**Comment by jechlin** [comment]
> handle comments in JSON

Oh - hrm, I didn't find that. One for the backlog!

**Comment by vojtechhabarta** [comment]
I haven't try monaco editor with comments in json but the same module is used by VS Code where comments in json are possible.

I found this: https://github.com/Microsoft/monaco-json/blob/master/src/monaco.d.ts but - there is `DiagnosticsOptions.allowComments`.

**Comment by jechlin** [comment]
yep... ok changed and pushed

**Comment by jechlin** [comment]
Had to revert because jackson didn't want to deserialize it... another time.

**Comment by vojtechhabarta** [comment]
It just need to be enabled in Jackson (feature 'ALLOW_COMMENTS').

**Comment by jechlin** [comment]
I couldn't find that for some reason. OK done!

**Comment by vojtechhabarta** [comment]
Closing, @jechlin thanks for your work on this issue 2 years ago 🙂
